mentat/embeddings.py

```diff
@@ -1,77 +1,124 @@
-
-    def exists(self, key: str) -> bool:
-        with self.conn as db:
-            cursor = db.execute("SELECT 1 FROM embeddings WHERE checksum=?", (key,))
-            return cursor.fetchone() is not None
+            output += response
+        return output
 
-    def __del__(self):
-        self.conn.close()
 
+class Collection:
+    _collection = None
 
-database = EmbeddingsDatabase()
+    def __init__(self, embedding_model: str):
+        self._collection = client.get_or_create_collection(
+            name=f"mentat-{embedding_model}",
+            embedding_function=MentatEmbeddingFunction(),  # type: ignore
+        )
+        self.migrate_old_db()
+
+    def exists(self, id: str) -> bool:
+        assert self._collection is not None, "Collection not initialized"
+        return len(self._collection.get(id)["ids"]) > 0
+
+    def add(self, checksums: list[str], texts: list[str]) -> None:
+        assert self._collection is not None, "Collection not initialized"
+        return self._collection.add(  # type: ignore
+            ids=checksums,
+            documents=texts,
+            metadatas=[{"active": False} for _ in checksums],
+        )
 
+    def query(self, prompt: str, checksums: list[str]) -> dict[str, float]:
+        assert self._collection is not None, "Collection not initialized"
 
-def _cosine_similarity(v1: list[float], v2: list[float]) -> float:
-    """Calculate the cosine similarity between two vectors."""
-    dot_product = np.dot(v1, v2)
-    norm_v1 = np.linalg.norm(v1)
-    norm_v2 = np.linalg.norm(v2)
-    return dot_product / (norm_v1 * norm_v2)  # pyright: ignore
+        self._collection.update(  # type: ignore
+            ids=checksums,
+            metadatas=[{"active": True} for _ in checksums],
+        )
+        results = self._collection.query(  # type: ignore
+            query_texts=[prompt],
+            where={"active": True},
+            n_results=len(checksums) + 1,
+        )
+        self._collection.update(  # type: ignore
+            ids=checksums,
+            metadatas=[{"active": False} for _ in checksums],
+        )
+        assert results["distances"], "Error calculating distances"
+        return {c: e for c, e in zip(results["ids"][0], results["distances"][0])}
+
+    def migrate_old_db(self):
+        """Temporary helper function to migrate sqlite3 to chromadb
+
+        Prior to January 2024, embeddings were fetched directly from the OpenAI API in
+        batches and saved to a db. We're currently using the same embeddings (ada-2) with
+        ChromaDB, so we might as well save the effort of re-fetching them. One drawback
+        is that ChromaDB saves the actual text, while our old schema did not, so migrated
+        records will have an empty documents field. This shouldn't be a problem. If it is,
+        we can just update the 'exists' method to require a non-empty "document" field.
+
+        TODO: erase this method/call after a few months
```

Comment:
this is a nice optimization, but I'm not sure the extra complexity is worth saving users a couple cents to rebuild. It might be unlikely, but some bug from migrated records being slightly different could be a big headache and not worth the risk

Reply:
Ok make sense. I was just looking at the size of my sqlite file but (1) mine is probably bigger than anyone's and (2) it is indeed really fast and cheap to reload.